### PR TITLE
Add bulk messaging support

### LINKED ISSUE #18: Add bulk messaging capability
Would you take a PR for adding bulk messaging support?

I recommend adding two options `threshold` and `maxDelay` like I used in [good-tcp](https://github.com/zefferus/good-tcp) which follows the pattern established by the official good plugins, such as [good-udp](https://github.com/hapijs/good-udp). These would inform the plugin if/how to buffer the messages before sending.

- `threshold` indicates how many messages to buffer before sending
- `maxDelay` tells the plugin the longest time it should hold a buffer before sending on next message (useful for programs with lower log frequency).

To maintain backward compatibility with existing plugin installations and make bulk transmission an opt-in feature, `threshold` would be set to `0`.

### PR DESCRIPTION
**Pull request by PizzaBrandon**
Resolves #18 based on feature details discussed there

### CODE REVIEW

File: test/index.js [hunk 1]
```diff
@@ -155,4 +207,108 @@ describe('GoodLoggly', () => {
...
+        expect(event[0]).to.equal(Merge({}, bulkResult[0], optionalFields));
+        expect(event[1]).to.equal(Merge({}, bulkResult[1], optionalFields));
+        expect(tags).to.be.a.function();
+        tags(null);
+      };
+
+      stream.pipe(reporter);
+      stream.push(bulkLogEventData[0]);
+      stream.push(bulkLogEventData[1]);
+      setTimeout(() => {
```
Comment (cjihrig) [inline test/index.js]: Can you use `setImmediate()` here instead.

File: lib/index.js [hunk 2]
```diff
@@ -37,7 +46,45 @@ class GoodLoggly extends Stream.Writable {
 
     data.timestamp = GoodLoggly.timeString(data.timestamp);
     data.msg = GoodLoggly.getMessage(data);
-    this._client.log(data, data.tags, callback);
+
+    this._buffer.push(data);
+
+    if (this._bufferReady) {
+      this._sendMessages((err) => {
+        this._bufferStart = null;
```
Comment (cjihrig) [inline lib/index.js]: `_bufferStart` should at least be defined in the constructor.

Comment (cjihrig) [inline lib/index.js]: This reset logic could be included in `_sendMessages()`, and you could just pass `callback` directly to `_sendMessages()`.

File: lib/index.js [hunk 3]
```diff
@@ -24,6 +24,15 @@ class GoodLoggly extends Stream.Writable {
       useTagHeader: false,
       tags: Array.isArray(config.tags) ? config.tags : []
     });
+
+    this._buffer = [];
+
+    // Send last messages on your way out
+    this.once('finish', () => {
+      if (this._buffer.length) {
+        this._sendMessages(() => {});
```
Comment (cjihrig) [inline lib/index.js]: You could make the callback to `_sendMessages()` optional so this would be unnecessary.

File: lib/index.js [hunk 4]
```diff
@@ -24,6 +24,15 @@ class GoodLoggly extends Stream.Writable {
       useTagHeader: false,
       tags: Array.isArray(config.tags) ? config.tags : []
     });
+
+    this._buffer = [];
+
+    // Send last messages on your way out
+    this.once('finish', () => {
+      if (this._buffer.length) {
```
Comment (cjihrig) [inline lib/index.js]: I'd also move this check into `_sendMessages()`, assuming that we don't want to allow empty arrays to be sent.

File: test/index.js (hunk 1 above)
Comment (PizzaBrandon) [inline test/index.js]: This one needs setTimeout to trigger exceeding the `maxDelay`. The other below can change to immediate.

**Comment by PizzaBrandon** [comment]
I have updated the PR with the changes as noted, with the one exception of the timeout.

File: lib/index.js [hunk 5]
```diff
@@ -37,7 +45,55 @@ class GoodLoggly extends Stream.Writable {
...
+      this._sendMessages(callback);
+    } else {
+      if (!this._bufferStart) {
+        this._bufferStart = Date.now();
+      }
+
+      setImmediate(callback);
+    }
+  }
+  get _bufferReady () {
```
Comment (arb) [inline lib/index.js]: I'm not a huge fan of `get`s. Can you make this just a regular function?

File: lib/index.js [hunk 6]
```diff
@@ -37,7 +45,55 @@ class GoodLoggly extends Stream.Writable {
...
+        // Backward-compatible behavior to do single log send
+        return this._client.log(this._buffer[0], this._buffer[0].tags, sendComplete);
+      }
+
+      // If using bulk, individual message tags can't be sent
+      return this._client.log(this._buffer, sendComplete);
+    }
+
+    // Fall-through behavior for empty buffer, ensuring async response
+    process.nextTick(() => { sendComplete(null); });
```
Comment (arb) [inline lib/index.js]: Unless you have a good reason, `setImmediate` is preferred to `process.nextTick()`

**Comment by PizzaBrandon** [comment]
Updated.

**Pull request merged**